diff --git a/scheduler/conf.c b/scheduler/conf.c
index bb6049b2c..943bac955 100644
--- a/scheduler/conf.c
+++ b/scheduler/conf.c
@@ -1127,8 +1127,8 @@ cupsdReadConfiguration(void)
 			     Group, 1, 1) < 0 ||
        cupsdCheckPermissions(StateDir, NULL, 0755, RunUser,
 			     Group, 1, 1) < 0 ||
-       cupsdCheckPermissions(StateDir, "certs", RunUser ? 0711 : 0511, User,
-			     SystemGroupIDs[0], 1, 1) < 0 ||
+       cupsdCheckPermissions(StateDir, "certs", 0711, RunUser,
+			     0, 1, 1) < 0 ||
        cupsdCheckPermissions(ServerRoot, NULL, 0755, RunUser,
 			     Group, 1, 0) < 0 ||
        cupsdCheckPermissions(ServerRoot, "ppd", 0755, RunUser,
