diff --git a/scheduler/cupsfilter.c b/scheduler/cupsfilter.c
index fa914d215..09df63234 100644
--- a/scheduler/cupsfilter.c
+++ b/scheduler/cupsfilter.c
@@ -921,7 +921,7 @@ exec_filters(mime_type_t   *srctype,	/* I - Source type */
 {
   int		i;			/* Looping var */
   const char	*argv[8],		/* Command-line arguments */
-		*envp[17],		/* Environment variables */
+		*envp[21],		/* Environment variables */
 		*temp;			/* Temporary string */
   char		*optstr,		/* Filter options */
 		content_type[1024],	/* CONTENT_TYPE */
@@ -929,10 +929,14 @@ exec_filters(mime_type_t   *srctype,	/* I - Source type */
 		cups_fontpath[1024],	/* CUPS_FONTPATH */
 		cups_serverbin[1024],	/* CUPS_SERVERBIN */
 		cups_serverroot[1024],	/* CUPS_SERVERROOT */
+		fontconfig_file[1024],	/* FONTCONFIG_FILE */
+		fontconfig_path[1024],	/* FONTCONFIG_PATH */
+		fontconfig_sysroot[1024], /* FONTCONFIG_SYSROOT */
 		final_content_type[1024] = "",
 					/* FINAL_CONTENT_TYPE */
 		lang[1024],		/* LANG */
 		path[1024],		/* PATH */
+		ld_library_path[2048],	/* LD_LIBRARY_PATH */
 		ppd[1024],		/* PPD */
 		printer_info[255],	/* PRINTER_INFO env variable */
 		printer_location[255],	/* PRINTER_LOCATION env variable */
@@ -1000,9 +1004,17 @@ exec_filters(mime_type_t   *srctype,	/* I - Source type */
            ServerBin);
   snprintf(cups_serverroot, sizeof(cups_serverroot), "CUPS_SERVERROOT=%s",
            ServerRoot);
+  snprintf(fontconfig_file, sizeof(fontconfig_file), "FONTCONFIG_FILE=%s",
+           getenv("FONTCONFIG_FILE"));
+  snprintf(fontconfig_path, sizeof(fontconfig_path), "FONTCONFIG_PATH=%s",
+           getenv("FONTCONFIG_PATH"));
+  snprintf(fontconfig_sysroot, sizeof(fontconfig_sysroot),
+	   "FONTCONFIG_SYSROOT=%s", getenv("FONTCONFIG_SYSROOT"));
   language = cupsLangDefault();
   snprintf(lang, sizeof(lang), "LANG=%s.UTF8", language->language);
   snprintf(path, sizeof(path), "PATH=%s", Path);
+  snprintf(ld_library_path, sizeof(ld_library_path), "LD_LIBRARY_PATH=%s",
+	   getenv("LD_LIBRARY_PATH"));
   if (ppdfile)
     snprintf(ppd, sizeof(ppd), "PPD=%s", ppdfile);
   else if ((temp = getenv("PPD")) != NULL)
@@ -1073,20 +1085,24 @@ exec_filters(mime_type_t   *srctype,	/* I - Source type */
   envp[5]  = cups_serverroot;
   envp[6]  = lang;
   envp[7]  = path;
-  envp[8]  = ppd;
-  envp[9]  = printer_info;
-  envp[10] = printer_location;
-  envp[11] = printer_name;
-  envp[12] = rip_max_cache;
-  envp[13] = userenv;
-  envp[14] = "CHARSET=utf-8";
+  envp[8]  = ld_library_path;
+  envp[9]  = ppd;
+  envp[10] = printer_info;
+  envp[11] = printer_location;
+  envp[12] = printer_name;
+  envp[13] = rip_max_cache;
+  envp[14] = userenv;
+  envp[15] = "CHARSET=utf-8";
+  envp[16] = fontconfig_file;
+  envp[17] = fontconfig_path;
+  envp[18] = fontconfig_sysroot;
   if (final_content_type[0])
   {
-    envp[15] = final_content_type;
-    envp[16] = NULL;
+    envp[19] = final_content_type;
+    envp[20] = NULL;
   }
   else
-    envp[15] = NULL;
+    envp[19] = NULL;
 
   for (i = 0; argv[i]; i ++)
     fprintf(stderr, "DEBUG: argv[%d]=\"%s\"\n", i, argv[i]);
@@ -1418,7 +1434,7 @@ read_cups_files_conf(
     cupsFileClose(fp);
   }
 
-  snprintf(line, sizeof(line), "%s/filter:" CUPS_BINDIR ":" CUPS_SBINDIR ":/bin:/usr/bin", ServerBin);
+  snprintf(line, sizeof(line), "%s/filter:%s:" CUPS_BINDIR ":" CUPS_SBINDIR ":/bin:/usr/bin", ServerBin, getenv("PATH"));
   set_string(&Path, line);
 
   return (0);
