--- scheduler/env.c.orig	2017-11-01 12:57:53.000000000 -0200
+++ scheduler/env.c	2017-11-16 15:59:30.831945067 -0200
@@ -212,8 +212,14 @@
   set_if_undefined("LD_PRELOAD", NULL);
   set_if_undefined("NLSPATH", NULL);
   if (find_env("PATH") < 0)
-    cupsdSetEnvf("PATH", "%s/filter:" CUPS_BINDIR ":" CUPS_SBINDIR
-			 ":/bin:/usr/bin", ServerBin);
+  {
+    char *value;
+    if ((value = getenv("PATH")) != NULL)
+      cupsdSetEnvf("PATH", "%s/filter:%s", ServerBin, value);
+    else
+      cupsdSetEnvf("PATH", "%s/filter:" CUPS_BINDIR ":" CUPS_SBINDIR
+		   ":/bin:/usr/bin", ServerBin);
+  }
   set_if_undefined("SERVER_ADMIN", ServerAdmin);
   set_if_undefined("SHLIB_PATH", NULL);
   set_if_undefined("SOFTWARE", CUPS_MINIMAL);
