--- scheduler/conf.c.orig	2017-11-01 12:57:53.000000000 -0200
+++ scheduler/conf.c	2017-11-16 15:46:46.570154424 -0200
@@ -3572,7 +3572,7 @@
       {
         int uid = atoi(value);
 
-	if (!uid)
+	if (0 && !uid)
 	{
 	  cupsdLogMessage(CUPSD_LOG_ERROR,
 	                  "Will not use User 0 as specified on line %d of %s "
@@ -3594,7 +3594,7 @@
 
 	if (p)
 	{
-	  if (!p->pw_uid)
+	  if (0 && !p->pw_uid)
 	  {
 	    cupsdLogMessage(CUPSD_LOG_ERROR,
 	                    "Will not use User %s (UID=0) as specified on line "
