--- cups/adminutil.c.orig	2019-12-17 19:55:55.311316920 -0300
+++ cups/adminutil.c	2019-12-17 22:00:05.353768661 -0300
@@ -1364,7 +1364,7 @@
   snprintf(name, namesize, "%s/cupsd.conf", cg->cups_serverroot);
   *remote = 0;
 
-#ifndef _WIN32
+#if 0
   if (!_cups_strcasecmp(host, "localhost") && !access(name, R_OK))
   {
    /*
@@ -1391,7 +1391,7 @@
       status = HTTP_STATUS_OK;
   }
   else
-#endif /* !_WIN32 */
+#endif /* 0 */
   {
    /*
     * Read cupsd.conf via a HTTP GET request...
